### PR TITLE
[ci-visibility] Fix negative duration in playwright steps

### LINKED ISSUE #4103: Timeout Errors not  Displayed in Datadog Dashboard Despite Successful Reporting with DD_TRACE_DEBUG
**Expected Behavior**

I expect that all test case errors, including timeouts, should be visible in the Datadog dashboard.

**Actual Behavior**

Timeout errors are not showing up in the Datadog dashboard, even though the traces are reported successfully, as indicated by the presence of `DD_TRACE_DEBUG` logs.

`DD_TRACE_DEBUG` output doesn't help much:

```
$ NODE_OPTIONS="-r dd-trace/ci/init" DD_ENV=ci DD_TRACE_DEBUG=true npx playwright test BO_Service.test.ts --project=$projectname

Extract from carrier: {"x-datadog-trace-id":"1176615104126108123","x-datadog-parent-id":"0000000000000000"}.
Running 1 test using 1 worker
Extract from carrier: {"x-datadog-trace-id":"617896122180169071","x-datadog-parent-id":"0000000000000000"}.

[1/1] [PROD] › BO_Service.test.ts:23:5 › UAT-TC-48: CC- Verify Get Customer Details with valid customerNumber
[PROD] › BO_Service.test.ts:23:5 › UAT-TC-48: CC- Verify Get Customer Details with valid customerNumber

Inject into carrier: {"x-datadog-trace-id":"3963407857172165997","x-datadog-parent-id":"3963407857172165997","x-datadog-sampling-priority":"1"}.

Flushing 0 metrics via HTTP
Flushing 0 metrics via HTTP
Flushing 0 metrics via HTTP

...

Encoding payload: [{"trace_id":"0893353fd6a3ed6f","span_id":"3bc39bf930498635","parent_id":"0000000000000000","name":"playwright.test","resource":"BO_Service.test.ts.UAT-TC-48: CC- Verify Get Customer Details with valid customerNumber","error":1,"meta":{"_dd.p.dm":"-4","service":"mena-prod","env":"ci","version":"1.0.0","...},"metrics":{"_dd.top_level":1,"_dd.rule_psr":1,"sampling.priority":1,...},"start":1708756519263642600,"duration":30021577393,"links":[],"service":"mena-prod","type":"test"},...]

...

Reporting results ...UAT-TC-48
Inject into carrier: {"x-datadog-trace-id":"7886921132486741761","x-datadog-parent-id":"7886921132486741761","x-datadog-sampling-priority":"1"}.

1) [PROD] › BO_Service.test.ts:23:5 › UAT-TC-48: CC- Verify Get Customer Details with valid customerNumber 
    "beforeAll" hook timeout of 30000ms exceeded.
...

Encoding payload: [{"trace_id":"6d73f8fbb964cf01","span_id":"6d73f8fbb964cf01","parent_id":"0000000000000000","name":"http.request","resource":"POST","error":0,"meta":{"_dd.p.tid":"65d98e4600000000","_dd.p.dm":"-0","service":"mena-prod","env":"ci","version":"1.0.0","...},"metrics":{"_dd.agent_psr":1,"_dd.top_level":1,"_dd.measured":1,...},"start":1708756550317011500,"duration":858751953,"links":[],"service":"mena-prod","type":"http"},...]

...

Request to the intake: {"path":"/api/v2/citestcycle","method":"POST","headers":{"Content-Type":"application/msgpack"},"timeout":15000,"url":"https://citestcycle-intake.datadoghq.com/"}
Response from the intake: {}

```

Environment

gitlab pipeline
Cloud Cl provider (Agentless)
playwright/test

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?

* A playwright's step can have a negative duration (`step.duration`) if it's not executed, as a result of a failing hook. This results in our intake rejecting the whole payload.
* In addition to fixing this, I've improved the way we report test errors in suites and sessions in playwright:
  * If a test suite fails as a result of a `beforeAll` or `afterAll` test, it will report an error.
  * If a test suite fails as a result of any of its tests failing, it'll report the test errors. 
  * Sessions will include an error message with the number of failed suites and errors if they fail.

### Motivation
Fixes #4103

### Plugin Checklist

- [x] Unit tests.
### Security 
Datadog employees:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

### CODE REVIEW

File: packages/datadog-instrumentations/src/playwright.js [hunk 1]
```diff
@@ -93,13 +99,56 @@ function getProjectsFromDispatcher (dispatcher) {
...
+  return hookError
+}
+
+function addErrorToTestSuite (testSuiteAbsolutePath, error) {
+  if (!testSuiteToErrors.has(testSuiteAbsolutePath)) {
+    testSuiteToErrors.set(testSuiteAbsolutePath, [error])
+  } else {
+    testSuiteToErrors.get(testSuiteAbsolutePath).push(error)
+  }
+}
```
Comment (romainkomorndatadog) [inline packages/datadog-instrumentations/src/playwright.js:141]: pet peeve nit: I think this reads more straightforward:
```suggestion
function addErrorToTestSuite (testSuiteAbsolutePath, error) {
  if (testSuiteToErrors.has(testSuiteAbsolutePath)) {
    testSuiteToErrors.get(testSuiteAbsolutePath).push(error)
  } else {
    testSuiteToErrors.set(testSuiteAbsolutePath, [error])
  }
}
```
"Just say no to `!`."

**Review by romainkomorndatadog — APPROVED** [review]

File: packages/datadog-instrumentations/src/playwright.js [hunk 2]
```diff
@@ -93,13 +99,56 @@ function getProjectsFromDispatcher (dispatcher) {
...
+      return _id === testProjectId
+    }
+    return name === testProjectId
+  })?.name
+}
+
+function formatTestHookError (error, hookType, isTimeout) {
+  let hookError = error
+  if (error) {
+    hookError = error
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js:126]: absolutely 🤦 . Thanks

File: packages/datadog-instrumentations/src/playwright.js (hunk 1 above)
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js:141]: agreed, thanks

File: packages/datadog-instrumentations/src/playwright.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: done in https://github.com/DataDog/dd-trace-js/pull/4113/commits/7d04ef231fb44cf3bc19b17806d17f9e0bdef2a8. I used your other comment to change another `!`

**Review by romainkomorndatadog — APPROVED** [review]

**Review by rochdev — APPROVED** [review]

**Pull request merged**